test/method.jl

```diff
@@ -0,0 +1,181 @@
+        @test_signature basic4() = 2
+    end
+
+    @testset "missing argnames" begin
+        @test_signature ma1(::Int32) = 2x
+        @test_signature ma2(::Int32, ::Bool) = 2x
+        @test_signature ma3(x, ::Int32) = 2x
+    end
+
+    @testset "Whereparams" begin
+        @test_signature f4(x::T, y::T) where T = 2x
+
+        @test_signature f5(x::S, y::T) where {S,T} = 2x
+        @test_signature f6(x::S, y::T) where {T,S} = 2x
+        @test_signature f7(x::S, y::T) where T where S = 2x
+    end
+
+    @testset "Whereparams with constraints" begin
+        @test_signature f8(x::S) where S<:Integer = 2x
+        @test_signature f9(x::S, y::S) where S<:Integer = 2x
+        @test_signature f10(x::S, y::T) where {S<:Integer, T<:Real} = 2x
+        @test_signature f11(x::S, y::Int64) where S<:Integer = 2x
+
+        @test_signature f12(x::S) where {S>:Integer} = 2x
+        @test_signature f13(x::S) where Integer<:S<:Number = 2x
+    end
+
+    @testset "Arg types with type-parameters" begin
+        @test_signature f14(x::Array{Int64, 1}) = 2x
+        @test_signature f15(x::Array{T, 1}) where T = 2x
+
+        @test_signature f16(x::Array{T, 1} where T<:Real) = 2x
+
+        # This is the same method as f16 (other than name), and one displaces the other
+        # but they have different method objects. And different (but equivelent) ASTd
+        # this generates something that should be the same as what `signature(f16)` does
+        # but with a gensym'd variable name
+        f16_alt(x::Array{<:Real, 1}) = 2x
+        f16_alt_sig = signature(first(methods(f16_alt)))
+        @test f16_alt_sig[:name] == :f16_alt
+        @test occursin(  # Hack to deal with gensymed name. Make it a string and use regex
+            r"^Expr\[:\(x::\(Array\{var\"(.*?)\", 1\} where var\"\1\" <: Real\)\)\]$",
+            string(f16_alt_sig[:args])
+        )
+    end
+
+    @testset "anon functions" begin
+        @test_signature (x) -> x  # no args
+        @test_signature (x)->2x
+
+        @test_signature ((::T) where T) -> 0   # Anonymous parameter
+
+
+    end
+
+    @testset "vararg" begin
+        @test_signature f17(xs::Vararg{Any, N} where N) = 2
+
+        # `f17_alt(xs...) = 2` lowers to the same method as `f18`
+        # but has a different AST according to `splitdef` so we can't us @test_signature
+        f17_alt(xs...) = 2
+        test_matches(
+            signature(only_method(f17_alt)),
+            Dict(
+                :name => :f17_alt,
+                :args => [:(xs::(Vararg{Any, N} where N))]
+            )
+        )
+
+        @test_signature f18(xs::Vararg{Int64, N} where N) = 2
+        @test_signature f19(x, xs::Vararg{Any, N} where N) = 2x
+    end
+
+    @testset "kwargs" begin  # We do not support them right now
+        #Following is broken:
+        #@test_signature kwargs17(x; y=3x) = 2x
+        kwargs17(x; y=3x) = 2x
+        # at least be sure we get the rest right:
```

Comment:
I think it may be preferable to error in this case? I can see people not familiar with this function getting burned that this only partially works

Reply:
I want this for the motivating use case in ChainRules.
I need the positional arguments.
But for keyword arguments, I only need to do `kwargs...`
which is something I can insert into the the `Dict` after `signature` has generated it.

Reply:
I'd be ok with some flag to have the function return partial results. If unset then it would throw an exception. 

Alternatively if you could set `def[:kwargs] = nothing` when kwargs are present then we at least are providing context that keywords are used by the method even though we don't know what they are. This would cause `combinedef` to fail unless explicitly removed.

Reply:
> Alternatively if you could set def[:kwargs] = nothing when kwargs are present then we at least are providing context that keywords are used by the method even though we don't know what they are. 

I think i like this idea.